bluesky/callbacks/stream.py

```diff
@@ -0,0 +1,181 @@
+                       'original_run_uid': doc['uid'],
+                       'time': ttime.time()},
+                      doc)
+        # Emit the start document for anyone subscribed to our Dispatcher
+        self.emit(DocumentNames.start, dict(md))
+
+    def descriptor(self, doc):
+        """Receive a descriptor and store for later reference"""
+        self._raw_descriptors[doc['uid']] = doc
+        super().descriptor(doc)
+
+    def send_event(self, doc):
+        """
+        Receive an event document
+
+        If we have not released a description yet, we can view the event
+        document and the last description we received from the raw RunEngine
+        and create a new updated description document. The event document is
+        also modified to have a new sequence number and description source so
+        that it is not confused with the raw data stream being emitted from the
+        RunEngine
+        """
+        # If we haven't described this configuration
+        # Send a new document to our subscribers
+        if doc['descriptor'] not in self._described:
+            prior_desc = self._raw_descriptors[doc['descriptor']]
+            # Create a new description document for the output of the stream
+            data_keys = dict()
+            # Parse the event document creating a new description. If the key
+            # existed in the original source description, just assume that it
+            # is the same type, units and shape. Otherwise do some
+            # investigation
+            for key, val in doc['data'].items():
+                # Described priorly
+                if key in prior_desc['data_keys']:
+                    key_desc = prior_desc['data_keys'][key]
+                # String key
+                elif isinstance(val, str):
+                    key_desc = {'dtype': 'string',
+                                'shape': []}
+                # Iterable
+                elif isinstance(val, Iterable):
+                    key_desc = {'dtype': 'array',
+                                'shape': np.shape(val)}
+                # Number
+                else:
+                    key_desc = {'dtype': 'number',
+                                'shape': []}
+                # Modify the source
+                key_desc['source'] = 'Stream: {}'.format(self.out_node.name
+                                                         or 'Unnamed Stream')
+                # Store in our new descriptor
+                data_keys[key] = key_desc
+            # Create our complete description document
+            desc = ChainMap({'uid': new_uid(), 'time': ttime.time(),
+                             'run_start': self._stream_start_uid,
+                             'object_keys': {'stream':
+                                             list(data_keys.keys())}},
+                            prior_desc)
+            # Store information about our descriptors
+            self._last_descriptor = dict(desc)
```

Comment:
I think instead of relying on recency to match a new event to its new descriptor, you should keep a mapping between original descriptor uids and their corresponding new descriptor documents.